### PR TITLE
[maintaince-aspnetcore] Add support Asp.Net Core 2.0

### PR DESCRIPTION
**Pull request by genusP**

### CODE REVIEW

**Comment by xuzhg** [comment]
@genusP Why do you think we should to target .Net Standard 2.0?

**Comment by ysmoradi** [comment]
@xuzhg I agree with you, but I'm really confused. /-:

I'm running [Microsoft.AspNet.OData 6](https://www.nuget.org/packages/Microsoft.AspNet.OData/) with  [ASP.NET Core Owin](https://www.nuget.org/packages/Microsoft.AspNetCore.Owin/) together on top of asp.net core 1.1 with full .net framework 4.7
It was working really fine, but when I updated my app to ASP.NET Core 2, I received following exception:

```
System.MissingMethodException: 'Method not found: 'System. IServiceProvider Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvi der(Microsoft.Extensions.DependencyInjection.IServiceCollection)
```

I cloned this repository and I updated project to .NET 4.6.1 (Minimum required .NET version to support .NET standard 2.0 with tooling 2). Then I updated Microsoft.Extensions.DependencyInjection to 2.0. **Surprisingly project gets build without any code modification.** Based on what I saw in project, this is not required, but why we're receiving this exception?

[Microsoft.AspNet.OData 6 nuget package](https://www.nuget.org/packages/Microsoft.AspNet.OData/) declares following in its dependencies: **Microsoft.Extensions.DependencyInjection (>= 1.0.0 && < 2.0.0)**

Based on current situation, **we can't use any .NET standard 2.0 compatible library which is using Microsoft.Extensions.DependencyInjection 2.0** such as Ef core, asp.net core etc with OData.

Could you please provide us some information about this? Thanks in advance.

**Comment by genusP** [comment]
@xuzhg AspNet Core 2.0 targeting only netsatndard2.0

**Comment by genusP** [comment]
@ysmoradi This is PR for Microsoft.AspNet**Core**.OData. You need clone branch maintaince-aspnetcore and build project in vNext folder.

**Comment by ysmoradi** [comment]
@genusP it's possible to run Microsoft.AspNet.OData 6 (the one developers use in ASP.NET/Katana projects) on top of asp.net core using owin integration. I prefer this approach because I can use stable version of Microsoft.AspNet.OData in asp.net core, instead of not tested Microsoft.AspNet**Core**.OData. Projects like swagger for odata work with Microsoft.AspNet.OData and I've lots of codes based on that package too. My approach is working very nice with asp.net core 1.1, but I stopped on ASP.NET Core 2 because of Microsoft.Extensions.DependencyInjection 2.0
To be clear, let's forget ASP.NET Core, imaging you'd like to use Microsoft.AspNet.OData with ASP.NET/Katana and **Entity framework core 2** you simply can't because of Microsoft.Extensions.DependencyInjection, odata works with version 1 of that and ef core needs version 2 of that. This is a **universal problem** in web api/odata no matter what you're doing and **your PR solves this completely**

**Comment by ysmoradi** [comment]
With help of "Microsoft.AspNet.OData **6.1**" which has "Microsoft.Extensions.DependencyInjection.Abstractions **(>= 1.0.0)**" dependency, and help of UseCustomContainerBuilder I should say that my own problem is now solved. I wrote the following:

```csharp
    public class ExtendedContainerBuilder : DefaultContainerBuilder, IContainerBuilder
    {
        public override IServiceProvider BuildContainer()
        {
            IServiceCollection services = (IServiceCollection)GetType().GetTypeInfo().BaseType.GetField(nameof(services), BindingFlags.NonPublic | BindingFlags.Instance).GetValue(this);

            return services.BuildServiceProvider(); // this code needs recompile against Microsoft.Extensions.DependencyInjection 2.0 That's why I say this PR is a good thing
        }
    }
```
My app is working fine on asp.net core 2. Thanks. (:
But if anyone else updates its Microsoft.Extensions.DependencyInjection to 2.0, her apps will simply fails to start. ):

**Comment by ysmoradi** [comment]
I found it! Finally.
https://github.com/aspnet/DependencyInjection/issues/550
You've to recompile this project against MS.DependencyInjection 2.0

**Comment by robward-ms** [comment]
@genusP - Looks like there is a merge conflict in the csproj from Sam's recent change. Can you pull the latest and rebase?

**Comment by genusP** [comment]
@robward-ms  conflict resolved

**Comment by ysmoradi** [comment]
I'm glad to see this PR is going to be merged. But something like that will happen to master branch? Microsoft.AspNet.OData 6 is using that library too and anybody who uses that can't use most .NET standard 2.0 libraries such as entity framework core 2.0. If you're not going to rebuilt that project against Microsoft.Extensions.DependencyInjection 2.0, maybe at least use following code which **works perfectly with both versions**:

```csharp
return (IServiceProvider)services.GetType().GetTypeInfo().Assembly.GetType("Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions").GetMethod("BuildServiceProvider", new[] { typeof(IServiceCollection) }).Invoke(null, new object[] { services });
```

Instead of https://github.com/OData/WebApi/blob/88c5f01f2ee2d685d873b2ab075909b7da6651f1/src/System.Web.OData/OData/DefaultContainerBuilder.cs#L81

File: vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj [hunk 1]
```diff
@@ -3,21 +3,28 @@
   <PropertyGroup>
     <Description>Microsoft.AspNetCore.OData</Description>
     <AssemblyTitle>Microsoft.AspNetCore.OData</AssemblyTitle>
-    <VersionPrefix>6.0.0-alpha1-170529</VersionPrefix>
-    <TargetFrameworks>netstandard1.6;net451;</TargetFrameworks>
+    <VersionPrefix>6.0.0-alpha1-1708161</VersionPrefix>
```
Comment (xuzhg) [inline vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj]: 180816 ?

File: vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj [hunk 2]
```diff
@@ -3,21 +3,28 @@
...
+    <TargetFrameworks>netstandard1.6;net451;netstandard2.0</TargetFrameworks>
     <AssemblyName>Microsoft.AspNetCore.OData</AssemblyName>
     <PackageId>Microsoft.AspNetCore.OData</PackageId>
     <RuntimeFrameworkVersion>1.0.4</RuntimeFrameworkVersion>
+    <Version>6.0.0-alpha1-1708161</Version>
```
Comment (xuzhg) [inline vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj]: 170816?

**Comment by xuzhg** [comment]
@ysmoradi For your above comment, did you file an issue? Or are you interested in creating a PR for us?

**Comment by xuzhg** [comment]
@genusP  Which APIs will be changed in your mentions?

**Comment by robward-ms** [comment]
@ysmoradi - Our plan is to use a version of this code in master but not merge it directly. I'll track your comment and make sure this is included when this code is included into master.

File: vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj (hunk 2 above)
Comment (genusP) [reply vNext/src/Microsoft.AspNetCore.OData/Microsoft.AspNetCore.OData.csproj]: year month day + build number of day

**Comment by genusP** [comment]
@xuzhg 
1. use DependencyInjection version 2, where the BuildProvider method was changed
2 changed ctor class ActionSelector
3 MediaTypeHeaderValue.MediaType property return type was changed

**Comment by xuzhg** [comment]
Merged at: b2710979e6ef13c0d0fc8cf32c2b32eecf420f29

Thanks your contribution.

**Comment by ysmoradi** [comment]
@xuzhg @robward-ms surprisingly I've not seen your last comments. Sorry about that. Is there anything I can do? Based on feedbacks I've received, I'm gonna create a PR. The issue was created by me before, I'll link PR to that issue too. Am I correct? Thanks for your attention

**Comment by robward-ms** [comment]
@ysmoradi - Yes, create a PR. I talked with Sam and I misunderstood your comment so ignore my response.

**Comment by justintoth** [comment]
7 months later, and still no support for the latest 2.0 packages in Microsoft.AspNet.OData. What am I missing (why was this only implemented for core???)

**Comment by genusP** [comment]
You can download OData for AspNet Core 2.0 by this link https://www.nuget.org/packages/Microsoft.AspNetCore.OData/7.0.0-beta2

**Comment by justintoth** [comment]
Yes but there is no support for non-core ASP.NET MVC is what I'm saying.

**Comment by xuzhg** [comment]
@justintoth 

Asp.Net Core version: https://www.nuget.org/packages/Microsoft.AspNetCore.OData/7.0.0-beta2
Asp.Net Classic version: https://www.nuget.org/packages/Microsoft.AspNet.OData/7.0.0-beta2

**Comment by robward-ms** [comment]
@justintoth - There is an open issue for AspNetCore with all of the history between this branch and the Nuget packages Sam has referenced: #939.

**Comment by justintoth** [comment]
Perfect, I'll take a look at the beta package for classic, thanks!

**Pull request closed without merging**